src/compute.rs

```diff
@@ -0,0 +1,215 @@
+                let end_index = path[path.len() - 1];
+                path_list.push(path);
+                visited[end_index as usize] = true;
+            }
+
+            // now we do bookkeeping for any found
+            // shortest paths.
+            for f in found_for_this_pathlen {
+                num_paths[f as usize] += 1;
+                num_paths[index] += 1;
+                total_path_length[f as usize] += pathlen;
+                total_path_length[index] += pathlen;
+                search_list.retain(|&x| x != f);
+                search_state[f as usize] = true;
+                if f == j {
+                    done = true;
+                }
+            }
+
+            // If no connection exists, stop searching for it.
+            if !touched {
+                search_list.retain(|&x| x != j);
+                search_state[j as usize] = true;
+                done = true
+            }
+
+            pathlen += 1;
+        }
+    }
+}
+
+/// this function is the thread task
+/// grabs next unprocessed node
+/// if no more nodes, exits
+/// returning betweenness, total path lengths, and num paths
+fn betweenness_task(
+    acounter: Arc<Mutex<usize>>,
+    aindices: Arc<Vec<Vec<GraphIndex>>>,
+) -> (Vec<u32>, Vec<u32>, Vec<u32>) {
+    let start = Instant::now();
+    let indices = &aindices;
+    let num_nodes = indices.len();
+
+    // each worker thread keeps its own cache of data
+    // these are returned when the thread finishes
+    // and then summed by the caller
+    let mut betweenness_count: Vec<u32> = vec![0; num_nodes];
+    let mut total_path_length: Vec<u32> = vec![0; num_nodes];
+    let mut num_paths: Vec<u32> = vec![0; num_nodes];
+
+    let mut finished = false;
+    while !finished {
+        let mut counter = acounter.lock().unwrap();
+        let index: usize = *counter;
+        *counter += 1;
+        drop(counter);
+        if index < num_nodes - 1 {
+            if index % 100 == 0 {
+                println!("node: {}, time: {:?}", index, start.elapsed());
+            }
+            betweenness_for_node(
+                index,
+                indices,
+                &mut betweenness_count,
+                &mut total_path_length,
+                &mut num_paths,
+            );
+        } else {
+            finished = true;
+        }
+    }
+    (betweenness_count, total_path_length, num_paths)
+}
+
+pub fn compute_betweenness(
+    indices: Vec<Vec<GraphIndex>>,
+    mut num_threads: usize,
+) -> (Vec<u32>, Vec<u32>, Vec<u32>) {
+    let start = Instant::now();
+    num_threads = num_threads.clamp(MIN_NUM_THREADS, MAX_NUM_THREADS);
+    println!("\ncompute: num_threads {:?}", num_threads);
+
+    let num_nodes = indices.len();
+
+    let mut betweenness_count: Vec<u32> = vec![0; num_nodes];
+    let mut total_path_length: Vec<u32> = vec![0; num_nodes];
+    let mut num_paths: Vec<u32> = vec![0; num_nodes];
+
+    let mut handles = Vec::new();
```

Comment:
The same functionality (or even a bit more) can be achieved using `JoinSet` which is dedicated to handle set of threads:
https://docs.rs/tokio/latest/tokio/task/struct.JoinSet.html

Reply:
This would require a dependency on `tokio`, not sure if that's the right move...

Reply:
I'm using `std::thread` here, not `tokio::task`.  Would there be any performance reason to use anything other than `std::thread`?

Reply:
I saw Niklas' comment after I wrote mine (page was not synced).